RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc

```diff
@@ -216,21 +248,200 @@ void CAHitQuadrupletGenerator::hitNtuplets(const IntermediateHitDoublets::Region
+  const TrackerTopology *tTopo=tTopoHand.product();	
+
+  const int numberOfHitsInNtuplet = 4;
+  std::vector<CACell::CAntuplet> foundQuadruplets;
+
+
+
+ int index =0;
+ for(const auto& regionLayerPairs: regionDoublets) {
+
+	 const TrackingRegion& region = regionLayerPairs.region();
+	 hitDoublets.clear();
+	 foundQuadruplets.clear();
+	  if (index == 0){   
+	  	createGraphStructure(layers, g);
+	  }
+	  else{  
+		clearGraphStructure(layers, g);
+	  }
+
+	  fillGraph(layers, g, hitDoublets,
+	            [&](const SeedingLayerSetsHits::SeedingLayer& inner,
+	                const SeedingLayerSetsHits::SeedingLayer& outer,
+	                std::vector<const HitDoublets *>& hitDoublets) {
+	      using namespace std::placeholders;
+	      auto found = std::find_if(regionLayerPairs.begin(), regionLayerPairs.end(),
+	                                std::bind(layerPairEqual, _1, inner.index(), outer.index()));
+	      if(found != regionLayerPairs.end()) {
+	        hitDoublets.emplace_back(&(found->doublets()));
+	        return true;
+	      }
+	      return false;
+	    });
+
+	CellularAutomaton ca(g);
+
+	ca.createAndConnectCells(hitDoublets, region, caThetaCut,
+			caPhiCut, caHardPtCut);
+
+	ca.evolve(numberOfHitsInNtuplet);
+
+	ca.findNtuplets(foundQuadruplets, numberOfHitsInNtuplet);
+
+
+	const QuantityDependsPtEval maxChi2Eval = maxChi2.evaluator(es);
+
+ 	// re-used thoughout, need to be vectors because of RZLine interface
+	std::array<float, 4> bc_r;
+	std::array<float, 4> bc_z;
+  	std::array<float, 4> bc_errZ2;
+  	std::array<GlobalPoint, 4> gps;
+  	std::array<GlobalError, 4> ges;
+  	std::array<bool, 4> barrels;
+	bool hasAlreadyPushedACandidate = false;
+	float selectedChi2 = std::numeric_limits<float>::max();
+
+ 	unsigned int numberOfFoundQuadruplets = foundQuadruplets.size();
+
+  	// Loop over quadruplets
+  	for (unsigned int quadId = 0; quadId < numberOfFoundQuadruplets; ++quadId)
+  	{
+
+    		auto isBarrel = [](const unsigned id) -> bool
+    		{
+      		return id == PixelSubdetector::PixelBarrel;
+    		};
+    		for(unsigned int i = 0; i< 3; ++i)
+    		{
+        		auto const& ahit = foundQuadruplets[quadId][i]->getInnerHit();
+        		gps[i] = ahit->globalPosition();
+        		ges[i] = ahit->globalPositionError();
+        		barrels[i] = isBarrel(ahit->geographicalId().subdetId());
+    		}
+
+    		auto const& ahit = foundQuadruplets[quadId][2]->getOuterHit();
+    		gps[3] = ahit->globalPosition();
+    		ges[3] = ahit->globalPositionError();
+    		barrels[3] = isBarrel(ahit->geographicalId().subdetId());
+    		if(caOnlyOneLastHitPerLayerFilter)
+   		{
+	 		unsigned int fourthLayerId = 0;
+  			unsigned int previousfourthLayerId = 0;
+ 	 		int subDetId = 0; 
+ 	 		int previousSubDetId = 0;
+	 		unsigned int sideId = 0; 
+ 	 		unsigned int previousSideId = 0;
+	 		std::array<unsigned int, 2> previousCellIds ={{0,0}};
+	  		bool isTheSameTriplet = false;
+			bool isTheSameFourthLayer = false;
```

Comment:
Could as many of these variable definitions as possible be moved where they're initialized with a sensible value, and changed to const if not modified? E.g.
```
const auto fourthLayerId = tTopo->layer(ahit->geographicalId());
```